### ISSUE TITLE
Callback on the initSession() function doesn't get called when app already installed

### ISSUE THREAD

**Issue by javidsheikh** [BranchMetrics/ios-branch-deep-linking-attribution#942]
When I launch the app from a universal link on a device where the app is already installed, the callback on the initSession() function doesn't get called. This is my `AppDelegate` code:

    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey : Any]? = nil) -> Bool {
        
        window = UIWindow(frame: UIScreen.main.bounds)
        
        func showHomeViewController() {
            
            //TODO: Remove for launch
            Branch.useTestBranchKey()

            Branch.getInstance().initSession(launchOptions: launchOptions) { (params, error) in
                
                guard error == nil, let parameters = params as? [String: AnyObject], let encodedReferralEvent = parameters[Branch.BranchKey.referralEvent.rawValue] as? NSString else {

                    DispatchQueue.main.async {
                        self.setRootViewController(viewController: self.produceRootViewController(referralEvent: nil))
                    }
                    
                    return
                }
                
                DispatchQueue.main.async {
                    let referralEvent = API.Users.Get.MyDecodedInvitationObject(base64EncodedString: encodedReferralEvent as String)
                    
                    self.setRootViewController(viewController: self.produceRootViewController(referralEvent: referralEvent))
                }
            }
        }
        
        #if DEBUG
        if EnvironmentPickerViewController.shouldShowEnvironmentPicker() {
            let envPicker = EnvironmentPickerViewController.newInstance()
            envPicker.completion = showHomeViewController
            window?.rootViewController = envPicker
            window?.makeKeyAndVisible()
        } else {
            showHomeViewController()
        }
        
        #else

            showHomeViewController()
        
        #endif

        return true
    }
    
    func setRootViewController(viewController: UIViewController) {
        window?.rootViewController = viewController
        window?.makeKeyAndVisible()
    }
    
    func produceRootViewController(referralEvent: ReferralEvent?) -> RootViewController {
        let vc = ControllerFactory.newController(instantiator: .storyboard, configurator: RootConfigurator())
        var destinationDS = vc.router?.dataStore
        destinationDS?.referralEvent = referralEvent
        return vc
    }
    
    func application(_ app: UIApplication, open url: URL, options: [UIApplication.OpenURLOptionsKey : Any] = [:]) -> Bool {
        Branch.getInstance().application(app, open: url, options: options)
        return true
    }
    
    func application(_ application: UIApplication, continue userActivity: NSUserActivity, restorationHandler: @escaping ([Any]?) -> Void) -> Bool {
        // handler for Universal Links
        Branch.getInstance().continue(userActivity)
        return true
    }
    
    func application(_ application: UIApplication, didReceiveRemoteNotification userInfo: [AnyHashable : Any], fetchCompletionHandler completionHandler: @escaping (UIBackgroundFetchResult) -> Void) {
        // handler for Push Notifications
        Branch.getInstance().handlePushNotification(userInfo)
    }

When I launch the app after installing it from the App Store, the callback is called. Any ideas why there is this difference in behaviour?

**Comment by echo-branch** [comment]
There's a known race condition with slow initialization, such as search ads or initializing something large before Branch.  Working on a fix. 

Workaround
https://github.com/BranchMetrics/ios-branch-deep-linking-attribution/blob/master/Branch-SDK/Branch-SDK/Branch.h#L683

**Comment by echo-branch** [comment]
0.29.0 should have a fix for this issue.

**Comment by javidsheikh** [comment]
@echo-branch thanks for your replies.
Still experiencing this problem. Using version 0.29.0 and have tried the suggested workaround too. Any other ideas?

**Comment by echo-branch** [comment]
Could you move the Branch init outside the controller init.  I suspect we're not around when the lifecycle call arrives.

**Comment by javidsheikh** [comment]
Still doesn't get called, even when I simplify the delegate code to this:

    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
      // if you are using the TEST key
      Branch.setUseTestBranchKey(true)
      // listener for Branch Deep Link data
      Branch.getInstance().initSession(launchOptions: launchOptions) { (params, error) in
        // do stuff with deep link data (nav to page, display content, etc)
        print(params as? [String: AnyObject] ?? {})
      }
      return true
    }

**Comment by doanduchieu** [comment]
> Still doesn't get called, even when I simplify the delegate code to this:
> 
> ```
> func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
>   // if you are using the TEST key
>   Branch.setUseTestBranchKey(true)
>   // listener for Branch Deep Link data
>   Branch.getInstance().initSession(launchOptions: launchOptions) { (params, error) in
>     // do stuff with deep link data (nav to page, display content, etc)
>     print(params as? [String: AnyObject] ?? {})
>   }
>   return true
> }
> ```

I have the same problem. Has it been fixed?

**Comment by echo-branch** [comment]
I haven't been able to repro this issue as described and will need more information about your app.

1.  Is there any setup before Branch init?  If we start late, we won't have a chance to see the link come in.
2.  Could you log when you init Branch, and when app calls continue user activity or open url?  init needs to finish before the other two calls.
3.  Are there any 3rd party SDKs intercepting calls to continue user activity or open url?

**Comment by doanduchieu** [comment]
Hi @echo-branch , steps to re produce are: 
1. Delete the app.
2. Install then run first -> It works in this case.
3. Exist the app.
4. Run app again -> It don't callback to `initSessionWithLaunchOptions`.

We was used `v 0.26.0` and it works fine.

> NSString* branchKey = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"branch_key"];
>     Branch *branch = [Branch getInstance:branchKey];
>     [branch delayInitToCheckForSearchAds];
>     [LinkManager sharedInstance].branch = branch;
>     [branch initSessionWithLaunchOptions:launchOptions andRegisterDeepLinkHandler:^(NSDictionary *params, NSError *error) {
>  if(error){
>         else {
>         }
>     }];

**Comment by echo-branch** [comment]
Researching.  

Could you check if disabling the search ad delay helps the situation?

**Comment by doanduchieu** [comment]
> Researching.
> 
> Could you check if disabling the search ad delay helps the situation?

`disabling the search ad delay helps the situation` it works in this case. Could you help me check about this problem?

**Comment by echo-branch** [comment]
I think I might have repro'd your situation.  I was missing it with logging turned on, but once I disabled all logging and debugging I was able to see similar behavior.  Still researching it.

**Comment by doanduchieu** [comment]
> I think I might have repro'd your situation. I was missing it with logging turned on, but once I disabled all logging and debugging I was able to see similar behavior. Still researching it.

Thanks @echo-branch, I'm waiting for a fix.

**Comment by doanduchieu** [comment]
Hi @echo-branch , Do you have any updates on this?

**Comment by echo-branch** [comment]
@doanduchieu Not yet.  Still researching.

**Comment by rpassis** [comment]
This is a pretty serious issue. We updated from `0.25.3` to `0.29.0` and started experiencing this change in behavior, which in our case led to the app failing to load as we were optimistically relying on that callback always running.

**Comment by rpassis** [comment]
FWIW this is where I think the regression is.

The multiple initialization calls above `@synchronized (self.asyncRequestCountLock)` increase the `asyncRequestCount` so when we hit the if branch that decides whether to call `[self initUserSessionAndCallCallback:YES];` or not there are inflight requests that haven't finished so the request count is not 0, and the SDK's user session is never initialized.

This certainly coincides with the diffs I'm seeing between 0.25.3 and 0.29.0 [here](https://github.com/BranchMetrics/ios-branch-deep-linking-attribution/compare/0.25.3...0.29.0#diff-fd80fc0a5541c21e3c1cba08a8501dbb)

This is reproducible pretty consistently for us while attached to the debugger so let me know if you need any additional information.

<img width="1161" alt="Screen Shot 2019-10-25 at 4 09 01 PM" src="https://user-images.githubusercontent.com/387079/67601264-eba67380-f741-11e9-8619-6173ec5b6237.png">

**Comment by echo-branch** [comment]
@rpassis There's a fix branch in progress.  The previous change was to mitigate a different race condition, but it fails in other scenarios.  There's a significant refactor of how we handle threading on startup.

**Comment by doanduchieu** [comment]
Hi @echo-branch, Is there any update from this issue? We are waiting a new version to remove UIWebView without this issue.

**Comment by javidsheikh** [comment]
@echo-branch any update?

**Comment by echo-branch** [comment]
@javidsheikh The fix is on the staging branch, QA is testing it.  I'm hoping to have a new release by mid next week.

**Comment by doanduchieu** [comment]
> @javidsheikh The fix is on the staging branch, QA is testing it. I'm hoping to have a new release by mid next week.

Hi @echo-branch , have you released a new version to fix this?

**Comment by echo-branch** [comment]
0.30.0 is released and includes general init threading improvements.

**Comment by echo-branch** [comment]
Should be fixed in recent releases

**Comment by carbonimax** [comment]
Still experiencing this problem using version 0.33.1. Only on the first launch and sometimes randomly after.

``` [[Branch getInstance] setRetryInterval:1.0];``` can help but it takes a long time.

**Comment by echo-branch** [comment]
@carbonimax could you open a new issue with some details about your setup?  It's likely you have a different issue.

**Comment by mohamedhamada1** [comment]
I'm facing the same issue

**Comment by echo-branch** [comment]
@mohamedhamada1 Could you open a new issue with some details about your setup?  It could be a different issue.

**Comment by elizhw** [comment]
This problem occasionally reappeared in 1.39.3
